### PR TITLE
Mark materials as constant on subdomain 

### PR DESCRIPTION
**Pull request by dschwen**
Refs #15417

### CODE REVIEW

File: modules/phase_field/include/materials/ConstantAnisotropicMobility.h [hunk 1]
```diff
@@ -24,15 +24,14 @@ class ConstantAnisotropicMobilityTempl : public Material
   ConstantAnisotropicMobilityTempl(const InputParameters & parameters);
 
 protected:
-  virtual void computeProperties(){};
-  virtual void initialSetup();
+  virtual void computeQpProperties();
 
   /// raw tensor values as passed in from the input file
-  std::vector<Real> _M_values;
+  std::vector<Real> _mobility_values;
```
Comment (aeslaughter) [inline modules/phase_field/include/materials/ConstantAnisotropicMobility.h:30]: :smile:

File: framework/src/materials/GenericConstant2DArray.C [hunk 2]
```diff
@@ -25,6 +25,7 @@ GenericConstant2DArray::validParams()
   params.declareControllable("prop_value");
   params.addClassDescription(
       "A material evaluating one material property in type of RealEigenMatrix");
+  params.set<MooseEnum>("constant_on") = "SUBDOMAIN";
```
Comment (aeslaughter) [inline framework/src/materials/GenericConstant2DArray.C:28]: Do these objects now only work with "SUBDOMAIN" or whatever you are setting here? If so, you should suppress the parameter so it cannot be changed.

File: framework/include/materials/MaterialData.h [hunk 3]
```diff
@@ -128,9 +128,8 @@ class MaterialData
     return haveProperty<T>(prop_name);
   }
 
-  ///@{
   /**
-   * Methods for retieving a MaterialProperty object
+   * @{ Methods for retieving a MaterialProperty object
```
Comment (aeslaughter) [inline framework/include/materials/MaterialData.h:132]: What is this doing? Is there some fancy new Doxygen style I don't know about?

File: modules/phase_field/include/materials/ConstantAnisotropicMobility.h (hunk 1 above)
Comment (aeslaughter) [inline modules/phase_field/include/materials/ConstantAnisotropicMobility.h:30]:
```suggestion
  const std::vector<Real> & _mobility_values;
```

File: modules/phase_field/include/materials/ConstantAnisotropicMobility.h [hunk 4]
```diff
@@ -24,15 +24,14 @@ class ConstantAnisotropicMobilityTempl : public Material
...
 
   /// Name of the mobility tensor material property
-  MaterialPropertyName _M_name;
-  GenericMaterialProperty<RealTensorValue, is_ad> & _M;
+  MaterialPropertyName _mobility_name;
```
Comment (aeslaughter) [inline modules/phase_field/include/materials/ConstantAnisotropicMobility.h:33]:
```suggestion
const  MaterialPropertyName & _mobility_name;
```

File: framework/include/materials/MaterialData.h (hunk 3 above)
Comment (dschwen) [reply framework/include/materials/MaterialData.h:132]: No, any other option would still work but jut be less efficient.

File: modules/phase_field/include/materials/ConstantAnisotropicMobility.h (hunk 1 above)
Comment (dschwen) [reply modules/phase_field/include/materials/ConstantAnisotropicMobility.h:30]: The old member name is using a symbol format that is reserved for implementation specific purposes (underscore followed by a capital letter).

File: framework/src/materials/GenericConstant2DArray.C (hunk 2 above)
Comment (dschwen) [reply framework/src/materials/GenericConstant2DArray.C:28]: No, any other option would still work but jut be less efficient.

**Comment by dschwen** [comment]
Ok, I see this requires specific methods to get implemented. Weird. Why not call the regular `computeQpProperties` but call it just once per subdomain. I'll circle back to this.

**Comment by dschwen** [comment]
Great, at least one test (`userobjects/layered_side_integral.flux_average`) exodiffs, and the only material it uses is `GenericConstantMaterial`.

At this point I'm not very surprised, as it looks like the `computeSubdomainProperties()` is **not tested at all**. It is possible there's a regression in that code path. 

The functionality was originally introduced in 65816258587ec95aa4484c3f4330ee515e1ba4f5 by @arovinelli . I am curious what the design decision here was to have a separate method instead of calling `computeQpProperties` once per domain (like it is called once per element if you set it to be constant on an element). The introduction of `computeSubdomainProperties` makes it problematic when inheriting from a material that is constant and making the derived class potentially non-constant. You have to add more boilerplate code in that case. With a simpler design you'd just have to check the `constant_on` parameter (respectively just change it's default).

**Comment by arovinelli** [comment]
> Great, at least one test (`userobjects/layered_side_integral.flux_average`) exodiffs, and the only material it uses is `GenericConstantMaterial`.
> 
> At this point I'm not very surprised, as it looks like the `computeSubdomainProperties()` is **not tested at all**. It is possible there's a regression in that code path.
> 
> The functionality was originally introduced in [6581625](https://github.com/idaholab/moose/commit/65816258587ec95aa4484c3f4330ee515e1ba4f5) by @arovinelli [36e28ef](https://github.com/idaholab/moose/commit/36e28efb1e046329a96480652b04b1fa5888311a) by @YaqiWang (sorry, there is a lot of churn in those files). I am curious what the design decision here was to have a separate method instead of calling `computeQpProperties` once per domain (like it is called once per element if you set it to be constant on an element). The introduction of `computeSubdomainProperties` makes it problematic when inheriting from a material that is constant and making the derived class potentially non-constant. You have to add more boilerplate code in that case. With a simpler design you'd just have to check the `constant_on` parameter (respectively just change it's default).

Daniel, you scared me for a second!

**Comment by dschwen** [comment]
> Daniel, you scared me for a second!

Sorry! 😊

File: framework/include/materials/MaterialData.h (hunk 3 above)
Comment (aeslaughter) [reply framework/include/materials/MaterialData.h:132]: Cool, I have only used the ///@{

**Comment by dschwen** [comment]
Down to a Moltres failure now. Yay.

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**